satpy/modifiers/angles.py

```diff
@@ -138,33 +138,40 @@ def _zarr_pattern(self, arg_hash, cache_version: Union[None, int, str] = None) -
 
     def __call__(self, *args, cache_dir: Optional[str] = None) -> Any:
         """Call the decorated function."""
-        new_args = self._sanitize_args_func(*args) if self._sanitize_args_func is not None else args
-        arg_hash = _hash_args(*new_args, unhashable_types=self._uncacheable_arg_types)
-        should_cache, cache_dir = self._get_should_cache_and_cache_dir(new_args, cache_dir)
+        sanitized_args = self._sanitize_args_func(*args) if self._sanitize_args_func is not None else args
+        should_cache, cache_dir = self._get_should_cache_and_cache_dir(sanitized_args, cache_dir)
+        if should_cache:
+            try:
+                arg_hash = _hash_args(*sanitized_args, unhashable_types=self._uncacheable_arg_types)
+            except TypeError as err:
+                warnings.warn("Cannot cache function because of unhashable argument: " + str(err), stacklevel=2)
+                should_cache = False
+
+        if not should_cache:
+            return self._func(*args)
+
         zarr_fn = self._zarr_pattern(arg_hash)
         zarr_format = os.path.join(cache_dir, zarr_fn)
         zarr_paths = glob(zarr_format.format("*"))
-        if not should_cache or not zarr_paths:
-            # use sanitized arguments if we are caching, otherwise use original arguments
-            args_to_use = new_args if should_cache else args
+
+        if not zarr_paths:
+            # use sanitized arguments
+            args_to_use = sanitized_args
             res = self._func(*args_to_use)
-            if should_cache and not zarr_paths:
-                self._warn_if_irregular_input_chunks(args, args_to_use)
-                self._cache_results(res, zarr_format)
+            self._warn_if_irregular_input_chunks(args, args_to_use)
+            self._cache_results(res, zarr_format)
+
         # if we did any caching, let's load from the zarr files
```

Comment:
@djhoese while refactoring this function a bit, I got confused by this: Why do we load from the zarr files just after we cached it to disk?

Reply:
1. So that we're caching immediately and using those cached values.
2. So that future calls to this function that result in the same cached results will use the same cached values and when `.compute()` is finally called there is no redundant computation. By that I mean if we generate these angles once but then continue to use the dask array and then a different product generates these angles and detects the cached value and uses the cached arrays, then when we finally do `dask.array.compute(composite1, composite2)` we would get one set of computations based on the original dask array and one set of computations based on the cached array even though they should be the same numbers.

Reply:
ah, ok, good. I might improve the comment to explain point 2, for future us.